ipa-core/src/protocol/ipa_prf/validation_protocol/proof_generation.rs

```diff
@@ -80,39 +81,82 @@ impl ProofBatch {
-        const SPL: usize = SmallProofGenerator::PROOF_LENGTH;
+        const FRF: usize = FirstProofGenerator::RECURSION_FACTOR;
+        const FLL: usize = FirstProofGenerator::LAGRANGE_LENGTH;
+        const FPL: usize = FirstProofGenerator::PROOF_LENGTH;
+        const CRF: usize = CompressedProofGenerator::RECURSION_FACTOR;
+        const CLL: usize = CompressedProofGenerator::LAGRANGE_LENGTH;
+        const CPL: usize = CompressedProofGenerator::PROOF_LENGTH;
 
         // precomputation for first proof
-        let first_denominator = CanonicalLagrangeDenominator::<Fp61BitPrime, LRF>::new();
-        let first_lagrange_table = LagrangeTable::<Fp61BitPrime, LRF, LLL>::from(first_denominator);
+        let first_denominator = CanonicalLagrangeDenominator::<Fp61BitPrime, FRF>::new();
+        let first_lagrange_table = LagrangeTable::<Fp61BitPrime, FRF, FLL>::from(first_denominator);
+
+        let table_left_0: [[Fp61BitPrime; FPL]; 8] = array::from_fn(|i| {
+            let mut result = [Fp61BitPrime::ZERO; FPL];
+            let u = &TABLE_LEFT[i];
+            result[0..4].copy_from_slice(u);
+            let (a, b) = result.split_at_mut(8);
+            b.copy_from_slice(&first_lagrange_table.eval(<&[_; 8]>::try_from(&*a).unwrap()));
+            result
+        });
+        let table_left_1: [[Fp61BitPrime; FPL]; 8] = array::from_fn(|i| {
+            let mut result = [Fp61BitPrime::ZERO; FPL];
+            let u = &TABLE_LEFT[i];
+            result[4..8].copy_from_slice(u);
+            let (a, b) = result.split_at_mut(8);
+            b.copy_from_slice(&first_lagrange_table.eval(<&[_; 8]>::try_from(&*a).unwrap()));
+            result
+        });
+        let table_right_0: [[Fp61BitPrime; FPL]; 8] = array::from_fn(|i| {
+            let mut result = [Fp61BitPrime::ZERO; FPL];
+            let v = &TABLE_RIGHT[i];
+            result[0..4].copy_from_slice(v);
+            let (a, b) = result.split_at_mut(8);
+            b.copy_from_slice(&first_lagrange_table.eval(<&[_; 8]>::try_from(&*a).unwrap()));
+            result
+        });
+        let table_right_1: [[Fp61BitPrime; FPL]; 8] = array::from_fn(|i| {
+            let mut result = [Fp61BitPrime::ZERO; FPL];
+            let v = &TABLE_RIGHT[i];
+            result[4..8].copy_from_slice(v);
+            let (a, b) = result.split_at_mut(8);
+            b.copy_from_slice(&first_lagrange_table.eval(<&[_; 8]>::try_from(&*a).unwrap()));
+            result
+        });
+
+        let first_proof = FirstProofGenerator::compute_proof(table_index_tuple_inputs.chunk_array::<2>().map(
+            |chunk| {
+                let [(u0, v0), (u1, v1)] = chunk;
+                let u = array::from_fn(|i| table_right_0[usize::from(u0)][i] + table_right_1[usize::from(u1)][i]);
+                let v = array::from_fn(|i| table_left_0[usize::from(v0)][i] + table_left_1[usize::from(v1)][i]);
+                (u, v)
+            },
+        ));
```

Comment:
This could be optimized by deferring the reductions in these additions to happen at the same time as the reductions for the multiply-accumulate in `compute_proof`.